opm/simulators/linalg/bda/openclKernels.hpp

```diff
@@ -366,96 +366,78 @@ namespace bda
-                                         __global const unsigned int *rowptr,
-                                         __local double *localSum,
-                                         __local double *z1,
-                                         __local double *z2){
+    inline const char* stdwell_apply_s = R"(
+    __kernel void stdwell_apply(__global const double *Cnnzs,
+                                __global const double *Dnnzs,
+                                __global const double *Bnnzs,
+                                __global const int *Ccols,
+                                __global const int *Bcols,
+                                __global const double *x,
+                                __global double *y,
+                                __global const int *toOrder,
+                                const unsigned int dim,
+                                const unsigned int dim_wells,
+                                __global const unsigned int *val_pointers,
+                                __local double *localSum,
+                                __local double *z1,
+                                __local double *z2){
         int wgId = get_group_id(0);
         int wiId = get_local_id(0);
-        int valSize = rowptr[wgId + 1] - rowptr[wgId];
-        int valsPerBlock = blnc*blnr;
+        int valSize = val_pointers[wgId + 1] - val_pointers[wgId];
+        int valsPerBlock = dim*dim_wells;
         int numActiveWorkItems = (32/valsPerBlock)*valsPerBlock;
         int numBlocksPerWarp = 32/valsPerBlock;
-        int c = wiId % blnc;
-        int r = (wiId/blnc) % blnr;
+        int c = wiId % dim;
+        int r = (wiId/dim) % dim_wells;
         double temp;
 
+        barrier(CLK_LOCAL_MEM_FENCE);
+
         localSum[wiId] = 0;
         if(wiId < numActiveWorkItems){
-            int b = wiId/valsPerBlock + rowptr[wgId];
-            while(b < valSize + rowptr[wgId]){
-                int colIdx = colsB[b];
-                localSum[wiId] += valsB[b*blnc*blnr + r*blnc + c]*x[colIdx*blnc + c];
+            int b = wiId/valsPerBlock + val_pointers[wgId];
+            while(b < valSize + val_pointers[wgId]){
+                int colIdx = toOrder[Bcols[b]];
+                localSum[wiId] += Bnnzs[b*dim*dim_wells + r*dim + c]*x[colIdx*dim + c];
                 b += numBlocksPerWarp;
             }
-        }
 
-        barrier(CLK_LOCAL_MEM_FENCE);
+            if(wiId < valsPerBlock){
+                localSum[wiId] += localSum[wiId + valsPerBlock];
+            }
 
-        int stride = valsPerBlock;
-        if(wiId < stride){
-            localSum[wiId] += localSum[wiId + stride];
-        }
+            b = wiId/valsPerBlock + val_pointers[wgId];
 
-        barrier(CLK_LOCAL_MEM_FENCE);
+            if(wiId < valsPerBlock){
+                if(c == 0 || c == 2) {localSum[wiId] += localSum[wiId + 2];}
+                if(c == 0 || c == 1) {localSum[wiId] += localSum[wiId + 1];}
```

Comment:
Are we assuming dim<=3? In that case we need to throw somewhere if this does not hold (we might have additional polymer or other).

Also, I do not quite understand how this works. Why do we process `c == 1` in the second if-clause? Doesn't that mean that we are reading (if `c == 0` we read from w+1 for `wid == w`) and writing (`c == 1` holds for `wid = w+1`) for/to the same location? Probably I am missing something here.

Reply:
in the case the we are reading values written by different threads, we might also need synchronization/barriers in-between.

Reply:
We assume `dim == 3`. We process `c == 1` because we want to add 3 consecutive values of the `localSum`vector. I've written it like this because it is written in a similar manner in the CUDA kernel and I wanted to test if something was wrong with my previous implementation. I think the previous implementation was correct. I'll run some tests to see if this holds and if the other implementation is faster.

Reply:
Maybe the CUDA code is wrong, too. Or I am (as often) seeing ghosts. This is what I think that we are computing (assuming that in "step 1" we use the original localSum[4], but that value is changed at the same time by the other thread!)

| c | 0 | 1 | 2 | 0 | 1 | 2 |
|---|---|---|---|----|---|----|
| wid | 3 | 4 | 5 | 6 | 7 | 8 |
|---|---|---|---|----|---|----|
| step 0 | localSum[3]+localSum[5] | | localSum[5] + localSum[7] |
| step 1 | localSum[3]+localSum[5]+ localSum[4] | localSum[4]+localSum[5] + localSum[7] |

If that assumption holds we seem to compute the right value for c==0, but the value of c==1 is (hopefully) not used. Where is the flaw in my thinking?

If the same assumptions on dim and dimw are done in the CUDA, please leave changing these for another PR.